package goby.acomms.protobuf;

enum SlotType
{
  SLOT_DATA = 1;
  SLOT_PING = 2;
  SLOT_REMUS_LBL = 3;
};

message Slot
{
  required int32 src = 1;
  optional int32 dest = 2 [ default = 0 ];
  optional uint32 rate = 3 [ default = 0 ];
  required SlotType type = 4;
  optional uint32 slot_seconds = 5;

  // "2011-Jan-22 12:08:06.887950"
  // which is produced by operator<< for boost::posix_time::ptime
  optional string last_heard_time = 6; 
}

enum MACType {
  MAC_NONE = 1; // no MAC 
  MAC_FIXED_DECENTRALIZED = 2; // decentralized time division multiple access
  MAC_AUTO_DECENTRALIZED = 3; // decentralized time division multiple access with peer discovery
  MAC_POLLED = 4; // centralized polling
};

message MACConfig
{
  required int32 modem_id = 1;
  optional MACType type = 2 [default = MAC_NONE];

  // only required for MAC_FIXED_DECENTRALIZED and MAC_POLLED, specify templates for MAC_AUTO_DECENTRALIZED below
  repeated Slot cycle = 3;
  
  // only required for MAC_AUTO_DECENTRALIZED - specify these in `cycle` (above) for other types
  optional uint32 rate = 4 [default = 0];
  optional uint32 slot_seconds = 5 [default = 15];
  optional uint32 expire_cycles = 6 [default = 30];
  
}
